blob/src/main/java/com/bazaarvoice/emodb/blob/core/DefaultBlobStore.java

```diff
@@ -299,17 +338,36 @@ private void copyTo(ByteBuffer buf, OutputStream out) throws IOException {
     }
 
     @Override
-    public void put(String tableName, String blobId, InputSupplier<? extends InputStream> in, Map<String,String> attributes) throws IOException {
+    public void put(String tableName, String blobId, InputSupplier<? extends InputStream> in, Map<String, String> attributes) throws IOException {
         checkLegalTableName(tableName);
         checkLegalBlobId(blobId);
         checkNotNull(in, "in");
         checkNotNull(attributes, "attributes");
 
         Table table = _tableDao.get(tableName);
 
+        StorageSummary summary = putBlobObject(table, blobId, in, attributes);
+
+        try {
+            _metadataProvider.writeMetadata(table, blobId, summary);
+        } catch (Throwable t) {
+            LOGGER.error("Failed to upload metadata for table: {}, blobId: {}, attempt to delete blob. Exception: {}", tableName, blobId, t.getMessage());
+
+            try {
+                _storageProvider.deleteObject(table, blobId);
+            } catch (Exception e1) {
+                LOGGER.error("Failed to delete blob for table: {}, blobId: {}. Inconsistency between blob and metadata storages. Exception: {}", tableName, blobId, e1.getMessage());
+                _metadataInconsistencyMeter.mark();
+            } finally {
+                Throwables.propagate(t);
+            }
+        }
+    }
+
+    private StorageSummary putBlobObject(Table table, String blobId, InputSupplier<? extends InputStream> in, Map<String, String> attributes) throws IOException {
```

Comment:
minor - nitpick - inconsistent naming - BlobObject and BlobData within the new changes.

Reply:
will fix